python/shotgun_model/shotgun_hierarchy_model.py

```diff
@@ -0,0 +1,847 @@
+        """
+        Returns True if parent has any children; otherwise returns False.
+
+        :param index: The index of the item being tested.
+        :type index: :class:`~PySide.QtCore.QModelIndex`
+        """
+
+        if not index.isValid():
+            return super(ShotgunHierarchyModel, self).hasChildren(index)
+
+        item = self.itemFromIndex(index)
+        item_data = get_sg_data(item)
+
+        if not item_data:
+            # could not get the item data, let the base class check
+            return super(ShotgunHierarchyModel, self).hasChildren(index)
+
+        # the nav data knows whether it has children
+        return item_data.get("has_children", False)
+
+    def fetchMore(self, index):
+        """
+        Returns True if there is more data available for parent; otherwise
+        returns False.
+
+        :param index: The index of the item being tested.
+        :type index: :class:`~PySide.QtCore.QModelIndex`
+        """
+
+        if not index.isValid():
+            return
+
+        item = self.itemFromIndex(index)
+        item_data = get_sg_data(item)
+
+        if not item_data:
+            return
+
+        path = item_data["url"]
+
+        # set the flag to prevent subsequent attempts to fetch more
+        item.setData(True, self.SG_ITEM_FETCHED_MORE)
+
+        # query the information for this item to populate its children.
+        # the slot for handling worker success will handle inserting the
+        # queried data into the tree.
+        logger.debug("Fetching more for item: %s" % (item.text(),))
+        self._query_hierarchy(
+            path,
+            self._seed_entity_field,
+            self._entity_fields
+        )
+
+    def canFetchMore(self, index):
+        """
+        Returns True if there is more data available for parent; otherwise
+        returns False.
+
+        :param index: The index of the item being tested.
+        :type index: :class:`~PySide.QtCore.QModelIndex`
+        """
+
+        if not index.isValid():
+            return False
+
+        # get the item and it's stored hierarchy data
+        item = self.itemFromIndex(index)
+
+        if item.data(self.SG_ITEM_FETCHED_MORE):
+            # more data has already been queried for this item
+            return False
+
+        item_data = get_sg_data(item)
+
+        if not item_data:
+            return False
+
+        # the number of existing child items
+        child_item_count = item.rowCount()
+
+        # we can fetch more if there are no children already and the item
+        # has children.
+        return child_item_count == 0 and item_data.get("has_children", False)
+
+    ############################################################################
+    # protected methods
+
+    def _create_item(self, data, parent=None, row=None):
+        """
+        Creates a model item given the supplied data and optional parent.
+
+        The supplied ``data`` corresponds to the results of a call to the
```

Comment:
For every method that is protected and meant to be subclassed, I suggest we provide a small story in the docstring, explaining when it would be useful to subclass this method. I tried to do that with the Shotgun model and I think it helps both maintain and explore the code.